### PR TITLE
Add the ability to benchmark multiple models concurrently

### PR DESCRIPTION
**Pull request by liu-cong**
This is useful for benchmarking multiple LoRA adapters. I used this to [benchmark the gateway](https://docs.google.com/document/d/11ALHEF-9yOaLdbHbDjBoTY6fzejoEKiSYHzWpWqe8ZY/edit#heading=h.bow7ai44ybg5): 
- Also fix the latency_throughput_curve.sh to parse non-integer request rate properly.
- Also added "errors" to the benchmark results.

### CODE REVIEW

File: benchmarks/benchmark/tools/profile-generator/container/latency_throughput_curve.sh [hunk 1]
```diff
@@ -26,13 +26,19 @@ fi
 PYTHON="python3"
 PYTHON_OPTS="benchmark_serving.py "
 for request_rate in $(echo $REQUEST_RATES | tr ',' ' '); do
+  echo "Benchmaking request rate: ${request_rate}"
   # TODO: Check if profile already exists, if so then skip
   timestamp=$(date +"%Y-%m-%d_%H-%M-%S")
   output_file="latency-profile-${timestamp}.txt"
-  PYTHON_OPTS="$PYTHON_OPTS --save-json-results --host=$IP   --port=$PORT   --model=$TOKENIZER --dataset=$PROMPT_DATASET_FILE --tokenizer=$TOKENIZER --request-rate=$request_rate --backend=$BACKEND --num-prompts=$((request_rate * $BENCHMARK_TIME_SECONDS)) --max-input-length=$INPUT_LENGTH --max-output-length=$OUTPUT_LENGTH"
+  num_prompts=$(awk "BEGIN {print int($request_rate * $BENCHMARK_TIME_SECONDS)}")
```
Comment (liu-cong) [inline benchmarks/benchmark/tools/profile-generator/container/latency_throughput_curve.sh:33]: This is needed for non-integer request rate, e.g., 5.5

Comment (achandrasekar) [reply benchmarks/benchmark/tools/profile-generator/container/latency_throughput_curve.sh:33]: Do fractional request rates work as expected?

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py [hunk 2]
```diff
@@ -550,23 +551,72 @@ def main(args: argparse.Namespace):
...
+      args.num_prompts,
+      args.max_input_length,
+      args.max_output_length,
+      tokenizer,
+      args.use_dummy_text,
+  )
+  benchmark_start_time = time.time()
+  args.start_datetime = datetime.fromtimestamp(benchmark_start_time)
+  
+  results = await asyncio.gather(
```
Comment (achandrasekar) [inline benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:598]: When we specify multiple models, do we split the prompts / load among multiple models or do we run the same benchmark against each of the models and then aggregate? If you have an example output, that would be good to see.

**Comment by achandrasekar** [comment]
Thanks for sending this out @liu-cong! Change looks good overall and is very useful. 

@Bslabe123 if you can take a deeper look and make sure the existing cases with jetstream / vllm continue to work as expected that would be great.

File: benchmarks/benchmark/tools/profile-generator/container/latency_throughput_curve.sh (hunk 1 above)
Comment (liu-cong) [reply benchmarks/benchmark/tools/profile-generator/container/latency_throughput_curve.sh:33]: yes they do!

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py (hunk 2 above)
Comment (liu-cong) [reply benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:598]: We run the same load for each model, and then aggregate. E.g., request_rate=1, models=model1,model2, each model will get 1 QPS (2 QPS in total). 

Here is an example:
Model 1 output:
```
{
    "metrics": {
        "num_prompts_attempted": 5,
        "num_prompts_succeeded": 5,
        "request_rate": 1.0,
        "server_metrics": {},
        "benchmark_time": 8.14672040939331,
        "throughput_rps": 0.6137439053677247,
        "throughput": 34.124161138445494,
        "total_output_token": 278,
        "output_tokens_per_min": 2047.4496683067296,
        "total_input_tokens": 323,
        "input_tokens_per_min": 2378.871377205301,
        "total_tokens": 601,
        "tokens_per_min": 4426.321045512031,
        "avg_per_token_latency": 0.037309607130645586,
        "median_per_token_latency": 0.040652228564750856,
        "sd_per_token_latency": 0.0188187392701372,
        "min_per_token_latency": 0.002216795581070023,
        "max_per_token_latency": 0.05539818523692436,
        "p90_per_token_latency": 0.053828057389431506,
        "p99_per_token_latency": 0.05524117245217507,
        "avg_latency": 3701.921606063843,
        "median_latency": 1666.7413711547852,
        "sd_latency": 3176.3076078854724,
        "min_latency": 441.14232063293457,
        "max_latency": 8143.533229827881,
        "p90_latency": 7654.078578948975,
        "p99_latency": 8094.58776473999,
        "avg_per_output_token_latency": 68.41942499056644,
        "median_per_output_token_latency": 66.9147253036499,
        "sd_per_output_token_latency": 2.844504262535786,
        "min_per_output_token_latency": 65.6736550792571,
        "max_per_output_token_latency": 73.52372010548909,
        "p90_per_output_token_latency": 71.89325491587321,
        "p99_per_output_token_latency": 73.3606735865275,
        "avg_input_len": 64.6,
        "median_input_len": 23.0,
        "sd_input_len": 69.70394536896745,
        "min_input_len": 6.0,
        "max_input_len": 193.0,
        "p90_input_len": 149.4,
        "p99_input_len": 188.64,
        "avg_output_len": 55.6,
        "median_output_len": 24.0,
        "sd_output_len": 48.47102227104355,
        "min_output_len": 6.0,
        "max_output_len": 124.0,
        "p90_output_len": 116.0,
        "p99_output_len": 123.2,
        "ClientConnectorError": 0,
        "TimeoutError": 0,
        "ContentTypeError": 0,
        "ClientOSError": 0,
        "ServerDisconnectedError": 0,
        "unknown_error": 0
    },
    "dimensions": {
        "date": "20241015-173455",
        "backend": "vllm",
        "model_id": "sql-lora",
        "tokenizer_id": "meta-llama/Llama-2-7b-hf"
    },
    "config": {
        "model": "sql-lora",
        "num_models": 22,
        "model_server": "vllm",
        "start_time": {
            "seconds": 1729013695,
            "nanos": 45083000
        }
    },
    "summary_stats": {
        "stats": [
            {
                "request_rate": 1.0,
                "request_latency": {
                    "mean": 3701.921606063843,
                    "median": 1666.7413711547852,
                    "sd": 3176.3076078854724,
                    "min": 441.14232063293457,
                    "max": 8143.533229827881,
                    "p90": 7654.078578948975,
                    "p99": 8094.58776473999
                },
                "throughput": {
                    "mean": 34.124161138445494
                },
                "input_length": {
                    "mean": 64.6,
                    "median": 23.0,
                    "sd": 69.70394536896745,
                    "min": 6.0,
                    "max": 193.0,
                    "p90": 149.4,
                    "p99": 188.64
                },
                "output_length": {
                    "mean": 55.6,
                    "median": 24.0,
                    "sd": 48.47102227104355,
                    "min": 6.0,
                    "max": 124.0,
                    "p90": 116.0,
                    "p99": 123.2
                },
                "tpot": {
                    "mean": 68.41942499056644,
                    "median": 66.9147253036499,
                    "sd": 2.844504262535786,
                    "min": 65.6736550792571,
                    "max": 73.52372010548909,
                    "p90": 71.89325491587321,
                    "p99": 73.3606735865275
                },
                "model_server_metrics": []
            }
        ]
    }
}
```

Model 2 output:

```
{
    "metrics": {
        "num_prompts_attempted": 5,
        "num_prompts_succeeded": 5,
        "request_rate": 1.0,
        "server_metrics": {},
        "benchmark_time": 8.143963813781738,
        "throughput_rps": 0.6139516474199799,
        "throughput": 33.89013093758289,
        "total_output_token": 276,
        "output_tokens_per_min": 2033.4078562549732,
        "total_input_tokens": 323,
        "input_tokens_per_min": 2379.6765853998418,
        "total_tokens": 599,
        "tokens_per_min": 4413.084441654815,
        "avg_per_token_latency": 0.036208494464183796,
        "median_per_token_latency": 0.03735055574556676,
        "sd_per_token_latency": 0.01824103647381315,
        "min_per_token_latency": 0.002729444647554177,
        "max_per_token_latency": 0.05614647536442197,
        "p90_per_token_latency": 0.052968586635842564,
        "p99_per_token_latency": 0.055828686491564035,
        "avg_latency": 3670.4988956451416,
        "median_latency": 1531.3727855682373,
        "sd_latency": 3178.363572438023,
        "min_latency": 543.1594848632812,
        "max_latency": 8141.2389278411865,
        "p90_latency": 7638.134431838989,
        "p99_latency": 8090.928478240967,
        "avg_per_output_token_latency": 69.98296237933931,
        "median_per_output_token_latency": 66.18728545995859,
        "sd_per_output_token_latency": 10.380289888522471,
        "min_per_output_token_latency": 62.66227960586548,
        "max_per_output_token_latency": 90.52658081054688,
        "p90_per_output_token_latency": 81.00853513498775,
        "p99_per_output_token_latency": 89.57477624299096,
        "avg_input_len": 64.6,
        "median_input_len": 23.0,
        "sd_input_len": 69.70394536896745,
        "min_input_len": 6.0,
        "max_input_len": 193.0,
        "p90_input_len": 149.4,
        "p99_input_len": 188.64,
        "avg_output_len": 55.2,
        "median_output_len": 24.0,
        "sd_output_len": 47.90991546642511,
        "min_output_len": 6.0,
        "max_output_len": 122.0,
        "p90_output_len": 114.8,
        "p99_output_len": 121.28,
        "ClientConnectorError": 0,
        "TimeoutError": 0,
        "ContentTypeError": 0,
        "ClientOSError": 0,
        "ServerDisconnectedError": 0,
        "unknown_error": 0
    },
    "dimensions": {
        "date": "20241015-173455",
        "backend": "vllm",
        "model_id": "tweet-summary",
        "tokenizer_id": "meta-llama/Llama-2-7b-hf"
    },
    "config": {
        "model": "tweet-summary",
        "num_models": 22,
        "model_server": "vllm",
        "start_time": {
            "seconds": 1729013695,
            "nanos": 45083000
        }
    },
    "summary_stats": {
        "stats": [
            {
                "request_rate": 1.0,
                "request_latency": {
                    "mean": 3670.4988956451416,
                    "median": 1531.3727855682373,
                    "sd": 3178.363572438023,
                    "min": 543.1594848632812,
                    "max": 8141.2389278411865,
                    "p90": 7638.134431838989,
                    "p99": 8090.928478240967
                },
                "throughput": {
                    "mean": 33.89013093758289
                },
                "input_length": {
                    "mean": 64.6,
                    "median": 23.0,
                    "sd": 69.70394536896745,
                    "min": 6.0,
                    "max": 193.0,
                    "p90": 149.4,
                    "p99": 188.64
                },
                "output_length": {
                    "mean": 55.2,
                    "median": 24.0,
                    "sd": 47.90991546642511,
                    "min": 6.0,
                    "max": 122.0,
                    "p90": 114.8,
                    "p99": 121.28
                },
                "tpot": {
                    "mean": 69.98296237933931,
                    "median": 66.18728545995859,
                    "sd": 10.380289888522471,
                    "min": 62.66227960586548,
                    "max": 90.52658081054688,
                    "p90": 81.00853513498775,
                    "p99": 89.57477624299096
                },
                "model_server_metrics": []
            }
        ]
    }
}
```

Aggregated output:
```
{
    "metrics": {
        "num_prompts_attempted": 10,
        "num_prompts_succeeded": 10,
        "request_rate": 1.0,
        "server_metrics": {},
        "benchmark_time": 8.14886736869812,
        "throughput_rps": 0.6135822039767486,
        "throughput": 67.98490820062374,
        "total_output_token": 554,
        "output_tokens_per_min": 4079.094492037424,
        "total_input_tokens": 646,
        "input_tokens_per_min": 4756.4892452277545,
        "total_tokens": 1200,
        "tokens_per_min": 8835.583737265179,
        "avg_per_token_latency": 0.0367590507974147,
        "median_per_token_latency": 0.03900139215515881,
        "sd_per_token_latency": 0.018540315314315815,
        "min_per_token_latency": 0.002216795581070023,
        "max_per_token_latency": 0.05614647536442197,
        "p90_per_token_latency": 0.055473014249674116,
        "p99_per_token_latency": 0.05607912925294719,
        "avg_latency": 3686.210250854492,
        "median_latency": 1599.0570783615112,
        "sd_latency": 3177.3746011351222,
        "min_latency": 441.14232063293457,
        "max_latency": 8143.533229827881,
        "p90_latency": 8141.468358039856,
        "p99_latency": 8143.326742649078,
        "avg_per_output_token_latency": 69.20119368495287,
        "median_per_output_token_latency": 66.63446697731789,
        "sd_per_output_token_latency": 7.650619166197971,
        "min_per_output_token_latency": 62.66227960586548,
        "max_per_output_token_latency": 90.52658081054688,
        "p90_per_output_token_latency": 75.22400617599486,
        "p99_per_output_token_latency": 88.99632334709167,
        "avg_input_len": 64.6,
        "median_input_len": 23.0,
        "sd_input_len": 69.70394536896747,
        "min_input_len": 6.0,
        "max_input_len": 193.0,
        "p90_input_len": 193.0,
        "p99_input_len": 193.0,
        "avg_output_len": 55.4,
        "median_output_len": 24.0,
        "sd_output_len": 48.19170053027803,
        "min_output_len": 6.0,
        "max_output_len": 124.0,
        "p90_output_len": 122.2,
        "p99_output_len": 123.82,
        "ClientConnectorError": 0,
        "TimeoutError": 0,
        "ContentTypeError": 0,
        "ClientOSError": 0,
        "unknown_error": 0,
        "ServerDisconnectedError": 0
    },
    "dimensions": {
        "date": "20241015-173455",
        "backend": "vllm",
        "model_id": "ALL-2-MODELS",
        "tokenizer_id": "meta-llama/Llama-2-7b-hf"
    },
    "config": {
        "model": "ALL-2-MODELS",
        "num_models": 22,
        "model_server": "vllm",
        "start_time": {
            "seconds": 1729013695,
            "nanos": 45083000
        }
    },
    "summary_stats": {
        "stats": [
            {
                "request_rate": 1.0,
                "request_latency": {
                    "mean": 3686.210250854492,
                    "median": 1599.0570783615112,
                    "sd": 3177.3746011351222,
                    "min": 441.14232063293457,
                    "max": 8143.533229827881,
                    "p90": 8141.468358039856,
                    "p99": 8143.326742649078
                },
                "throughput": {
                    "mean": 67.98490820062374
                },
                "input_length": {
                    "mean": 64.6,
                    "median": 23.0,
                    "sd": 69.70394536896747,
                    "min": 6.0,
                    "max": 193.0,
                    "p90": 193.0,
                    "p99": 193.0
                },
                "output_length": {
                    "mean": 55.4,
                    "median": 24.0,
                    "sd": 48.19170053027803,
                    "min": 6.0,
                    "max": 124.0,
                    "p90": 122.2,
                    "p99": 123.82
                },
                "tpot": {
                    "mean": 69.20119368495287,
                    "median": 66.63446697731789,
                    "sd": 7.650619166197971,
                    "min": 62.66227960586548,
                    "max": 90.52658081054688,
                    "p90": 75.22400617599486,
                    "p99": 88.99632334709167
                },
                "model_server_metrics": []
            }
        ]
    }
}
```

**Comment by liu-cong** [comment]
> /hold I am testing the terraform changes

I also tested terraform, looks good to me

**Review by Bslabe123 — APPROVED** [review]

**Review by achandrasekar — APPROVED** [review]

**Pull request merged**